src/therapy/database/database.py

```diff
@@ -0,0 +1,355 @@
+"""Provide abstract Database class and relevant tools for database initialization."""
+import abc
+import sys
+from enum import Enum
+from os import environ
+from pathlib import Path
+from typing import Dict, Generator, List, Optional, Set, Union
+
+import click
+
+from therapy.schemas import RecordType, RefType, SourceMeta, SourceName
+
+
+class DatabaseError(Exception):
+    """Create custom class for handling database exceptions"""
+
+
+class DatabaseInitializationError(DatabaseError):
+    """Create custom exception for errors during DB connection initialization."""
+
+
+class DatabaseReadError(DatabaseError):
+    """Create custom exception for lookup/read errors"""
+
+
+class DatabaseWriteError(DatabaseError):
+    """Create custom exception for write errors"""
+
+
+class AbstractDatabase(abc.ABC):
+    """Define the database interface. This class should never be called directly by a
+    user, but should be used as the parent class for all concrete database
+    implementations.
+    """
+
+    @abc.abstractmethod
+    def __init__(self, db_url: Optional[str] = None, **db_args) -> None:
+        """Initialize database instance.
+
+        Generally, implementing classes should be able to construct a connection by
+        something like a libpq URL or server endpoint URL. Any additional arguments or
+        DB-specific parameters can be passed as keywords.
+
+        :param db_url: address/connection description for database
+        :param db_args: any DB implementation-specific parameters
+        :raise DatabaseInitializationError: if initial setup fails
+        """
+
+    @abc.abstractmethod
+    def list_tables(self) -> List[str]:
+        """Return names of tables in database.
+
+        :return: Table names in database
+        """
+
+    @staticmethod
+    def _check_delete_okay() -> bool:
+        """Check that environmental conditions permit DB deletion, and require
+        confirmation.
+
+        :raise DatabaseWriteException: if skip confirmation variable is set -- manual
+        approval is required.
+        """
+        if environ.get(AWS_ENV_VAR_NAME, "") == AwsEnvName.PRODUCTION:
+            if environ.get(SKIP_AWS_DB_ENV_NAME, "") == "true":
```

Comment:
We could prob just stick with `None` as the default 

 ```suggestion
        if environ.get(AWS_ENV_VAR_NAME) == AwsEnvName.PRODUCTION:
            if environ.get(SKIP_AWS_DB_ENV_NAME) == "true":
```